File: algorithms/2_medium/C/01238_circular_permutation_in_binary_representation/gray_code.c
static inline int gray_to_bin(int g)
{
    int res = g;
    while (g != 0) {
        g >>= 1;
        res ^= g;
    }
    return res;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* circularPermutation(int n, int start, int* return_len)
{
    assert(n >= 1 && n <= 16);
    const int cnt = 1 << n;
    int* res = malloc((size_t)cnt * sizeof(*res));
    assert(res != NULL);
    *return_len = cnt;
    int curr = gray_to_bin(start);

    for (int i = 0; i < cnt; i+= 1) {
        res[i] = curr ^ (curr >> 1);
        curr = (curr + 1) % cnt;
    }

    return res;
}

File: algorithms/2_medium/C/01238_circular_permutation_in_binary_representation/gray_code.rs
impl Solution {
    #[inline(always)]
    fn gray_to_bin(mut g: i32) -> i32 {
        let mut res = g;
        while g != 0 {
            g >>= 1;
            res ^= g;
        }
        res
    }

    pub fn circular_permutation(n: i32, start: i32) -> Vec<i32> {
        assert!(n >= 1 && n <= 16);
        let cnt = 2_i32.pow(n as u32);
        let mut res = Vec::with_capacity(cnt as usize);
        let mut curr = Self::gray_to_bin(start);

        for _ in 0..cnt {
            res.push(curr ^ (curr >> 1));
            curr = (curr + 1) % cnt;
        }

        res
    }
}
